### ISSUE TITLE
Remove space between clips

### ISSUE THREAD

**Issue by JC3** [audacity/audacity#3924]
### Your idea

For the love of god, please add a "remove space between clips" menu item that removes gaps between clips.

Dragging the clips together is a no-go if there's more than a handful of them. The go-to workaround of using "Truncate Silence" doesn't actually work if your clips contain silence, or they contain sound < -100dB (the unnecessary lower limit of the Truncate Silence text box), or they are closer than 0.1 seconds (another unnecessary lower limit in the Truncate Silence) dialog. Plus it's slow since it scans the audio, which is silly since Audacity already knows where clip boundaries are. 

This is such a basic feature.

### Problem to be solved

Removing gaps between clips.

### Prior art

Literally every application (audio or not) that works with clips of anything has the ability to remove the space between them. Every single one.

**Comment by JC3** [comment]
Related: https://github.com/audacity/audacity/issues/3933

**Comment by JC3** [comment]
Related: https://github.com/audacity/audacity/issues/3934

**Comment by SteveDaulton** [comment]
> Please add a "remove space between clips" menu item that removes gaps between clips.

That's been available for years. Use "Truncate Silence"  (set to "Truncate" rather than "Compress"). See: https://manual.audacityteam.org/man/truncate_silence.html

Note that in Audacity 3.1.0 it is necessary to enable:
"Preferences > Track Behaviors -> Editing a clip can move other clips".

**Comment by JC3** [comment]
> That's been available for years. Use "Truncate Silence" (set to "Truncate" rather than "Compress"). See: https://manual.audacityteam.org/man/truncate_silence.html

As I stated already:

> The go-to workaround of using "Truncate Silence" doesn't actually work if your clips contain silence, or they contain sound < -100dB (an unnecessary lower limit in the Truncate Silence dialog), or they are closer than 0.1 seconds (another unnecessary lower limit in the Truncate Silence dialog). Plus it's slow since it scans the audio, which is not necessary for this use since Audacity already knows where clip boundaries are.

**Comment by JohnColket** [comment]
FYI: Forum user el_gallo also requested this feature here: https://forum.audacityteam.org/viewtopic.php?p=464263#p464263

**Comment by tondeaf** [comment]
This isn't about removing silence FROM clips, it's about removing EMPTY SPACE BETWEEN clips.

**Comment by djhend** [comment]
I could really use this feature at the moment. I've received the 14x32 track wav files from a concert recording and lining them all up is proving very time consuming. Any magic ideas? Running version 3.3.2 Build from May 5, 2023

**Comment by unfor19** [comment]
@JC3 I've created this Python package - [unfor19/audacity-scripting](https://github.com/unfor19/audacity-scripting).

You can install it with:
```bash
python -m pip install audacity-scripting
```

Then run the CLI command:
```bash
audacity_scripting clean-spaces --path "/path/to/my_project.aup3"
```

The above command copies the original file and removes spaces (gaps) between audio clips from the copied file.

Let me know if it works for you. Feel free to offer suggestions on how to improve this tool.

**Comment by SteveDaulton** [comment]
[unfor19](https://github.com/unfor19) How do you work around https://github.com/audacity/audacity/issues/4220 ?

**Comment by unfor19** [comment]
Hi @SteveDaulton, first, I'd like to say thank you for Audacity and the well-documented process for building it locally; it was super easy.

When I use the `GetInfo`, I round everything to `5`, see:
- [init Clilp](https://github.com/unfor19/audacity-scripting/blob/982085cc547ef893f0b89f17628a7f22879132b6/src/audacity_scripting/bridge/clip.py#L10-L18) 
- [calculate clip start time](https://github.com/unfor19/audacity-scripting/blob/982085cc547ef893f0b89f17628a7f22879132b6/src/audacity_scripting/bridge/clip.py#L68-L72)

I had some cases where `start` and `end` were opposing for some clips, so I filtered them out:

[source](https://github.com/unfor19/audacity-scripting/blob/982085cc547ef893f0b89f17628a7f22879132b6/src/audacity_scripting/bridge/clip.py#L69-L75)

`python
[            for raw_clip in clips_sorted:
                if raw_clip['start'] >= 0.0 and raw_clip['end'] > 0.0:
                    raw_clip['_id'] = "" + \
                        str(raw_clip['track']) + \
                        str(round(raw_clip['start'], 5))](https://github.com/unfor19/audacity-scripting/blob/982085cc547ef893f0b89f17628a7f22879132b6/src/audacity_scripting/bridge/clip.py#L69-L75)
`

I [calculate new positions](https://github.com/unfor19/audacity-scripting/blob/982085cc547ef893f0b89f17628a7f22879132b6/src/audacity_scripting/bridge/wrappers.py#L100-L123) of clips in a track with gaps by calculating their duration; After I know the duration of the first clip in the track, I add the other clips to the end of it according to their duration.

When it comes to precision, I did see some gaps when I used `round(, 4)`, so I used `round(, 5)`, and now everything works perfectly.

I hope that answered your question; I'm not sure it did, as I didn't experience too many "precision issues."

**Comment by SteveDaulton** [comment]
An alternative solution using a "Nyquist Macro" plug-in: https://audionyq.com/remove-space-between-clips/

**Comment by mppooks** [comment]
> ### Your idea
> Please add a "remove space between clips" menu item that removes gaps between clips.
> 
> Dragging the clips together is a no-go if there's more than a handful of them. The go-to workaround of using "Truncate Silence" doesn't actually work if your clips contain silence, or they contain sound < -100dB (an unnecessary lower limit in the Truncate Silence dialog), or they are closer than 0.1 seconds (another unnecessary lower limit in the Truncate Silence dialog). Plus it's slow since it scans the audio, which is not necessary for this use since Audacity already knows where clip boundaries are.
> 
> This is a very basic feature.
> 
> Generally speaking, Audacity fails in use cases that involve labelling areas to _keep_ rather than labelling areas to _cut_ during editing (there is for example "Split Delete" but no direct "Split Trim" or whatever you'd call it).
> 
> ### Problem to be solved
> Removing gaps between clips.
> 
> ### Prior art
> Literally every application (audio or not) that works with clips of anything has the ability to remove the space between them.
> 
> ### Additional context
> _No response_

I agree this would be a great feature. "Truncate Silence" does not do the job. The workaround is to "Fit Project to Width" and select everything after the gap, cut and paste it closer. But it feels very risky to do this.

**Comment by d0ugparker** [comment]
Mods please consider editing the title to show @SteveDaulton's excellent Nyquist macro plug-in answers the OP's needs.

**Comment by LWinterberg** [comment]
This won't be needed for Audacity 4 as it has proper ripple editing
